File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/preferences/AthanNumericPreference.java
package com.byagowi.persiancalendar.view.preferences;

import android.content.Context;
import android.content.Intent;
import android.util.AttributeSet;

import com.byagowi.persiancalendar.Constants;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.preference.EditTextPreference;

/**
 * Created by ebraminio on 2/16/16.
 */
public class AthanNumericPreference extends EditTextPreference {

    public AthanNumericPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public AthanNumericPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public AthanNumericPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AthanNumericPreference(Context context) {
        super(context);
    }

    private Double mDouble;

    // http://stackoverflow.com/a/10848393
    @Override
    public void setText(String text) {
        final boolean wasBlocking = shouldDisableDependents();
        mDouble = parseDouble(text);
        persistString(mDouble != null ? mDouble.toString() : null);
        final boolean isBlocking = shouldDisableDependents();
        if (isBlocking != wasBlocking) notifyDependencyChange(isBlocking);
        LocalBroadcastManager.getInstance(getContext())
                .sendBroadcast(new Intent(Constants.LOCAL_INTENT_UPDATE_PREFERENCE));
    }

    @Override
    public String getText() {
        return mDouble != null ? mDouble.toString() : null;
    }

    private Double parseDouble(String text) {
        try {
            return Double.parseDouble(text);
        } catch (NumberFormatException | NullPointerException e) {
            return null;
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/preferences/AthanNumericPreference.kt
package com.byagowi.persiancalendar.view.preferences

import android.content.Context
import android.content.Intent
import android.util.AttributeSet
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.EditTextPreference
import com.byagowi.persiancalendar.Constants

/**
 * Created by ebraminio on 2/16/16.
 */
class AthanNumericPreference : EditTextPreference {

  private var mDouble: Double? = null

  constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {}

  constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

  constructor(context: Context) : super(context) {}

  // http://stackoverflow.com/a/10848393
  override fun setText(text: String) {
    val wasBlocking = shouldDisableDependents()
    mDouble = parseDouble(text)
    persistString(mDouble?.toString())
    val isBlocking = shouldDisableDependents()
    if (isBlocking != wasBlocking) notifyDependencyChange(isBlocking)
    LocalBroadcastManager.getInstance(context)
        .sendBroadcast(Intent(Constants.LOCAL_INTENT_UPDATE_PREFERENCE))
  }

  override fun getText(): String? = if (mDouble != null) mDouble.toString() else null

  private fun parseDouble(text: String): Double? {
    return try {
      java.lang.Double.parseDouble(text)
    } catch (e: NumberFormatException) {
      null
    } catch (e: NullPointerException) {
      null
    }

  }
}
